tests/node.ts

```diff
@@ -0,0 +1,214 @@
+    config: {
+      transport: {
+        HoprConnect: {
+          bootstrapServers: bootstrapAddress ? [bootstrapAddress] : [],
+          // simulates a NAT
+          // DO NOT use this in production
+          __noDirectConnections: noDirectConnections,
+          __noWebRTCUpgrade: noWebRTCUpgrade
+        }
+      },
+      peerDiscovery: {
+        autoDial: false
+      }
+    },
+    dialer: {
+      // Temporary fix
+      addressSorter: (ma: Multiaddr) => ma
+    }
+  })
+
+  await node.start()
+  console.log(`node started`)
+  return node
+}
+
+function handleProtocol(node: LibP2P) {
+  node.handle(TEST_PROTOCOL, (struct: Handler) => {
+    pipe(
+      struct.stream.source,
+      (source: Stream['source']) => {
+        return (async function* () {
+          for await (const msg of source) {
+            const decoded = new TextDecoder().decode(msg.slice())
+
+            console.log(`Received message <${decoded}>`)
+
+            yield new TextEncoder().encode(`Echoing <${decoded}>`)
+          }
+        })()
+      },
+      struct.stream.sink
+    )
+  })
+}
+
+type CmdDef =
+  | {
+      cmd: 'wait'
+      waitForSecs: number
+    }
+  | {
+      cmd: 'dial'
+      targetIdentityName: string
+      targetPort: number
+    }
+  | {
+      cmd: 'msg'
+      msg: string
+      targetIdentityName: string
+      relayIdentityName: string
+    }
+
+async function executeCommands({ node, cmds }: { node: LibP2P; cmds: CmdDef[] }) {
+  for (const cmdDef of cmds) {
+    switch (cmdDef.cmd) {
+      case 'wait': {
+        console.log(`waiting ${cmdDef.waitForSecs} secs`)
+        await new Promise((resolve) => setTimeout(resolve, durations.seconds(cmdDef.waitForSecs)))
+        console.log(`finished waiting`)
+        break
+      }
+      case 'dial': {
+        const targetPeerId = await peerIdForIdentity(cmdDef.targetIdentityName)
+        const targetAddress = new Multiaddr(`/ip4/127.0.0.1/tcp/${cmdDef.targetPort}/p2p/${targetPeerId.toB58String()}`)
+        console.log(`dialing ${cmdDef.targetIdentityName}`)
+        await node.dial(targetAddress)
+        console.log(`dialed`)
+        break
+      }
+      case 'msg': {
+        const targetPeerId = await peerIdForIdentity(cmdDef.targetIdentityName)
+        const relayPeerId = await peerIdForIdentity(cmdDef.relayIdentityName)
+        //@ts-ignore
+        let conn: Handler
+
+        console.log(`msg: dialing ${cmdDef.targetIdentityName} though relay ${cmdDef.relayIdentityName}`)
+        conn = await node.dialProtocol(
+          new Multiaddr(`/p2p/${relayPeerId}/p2p-circuit/p2p/${targetPeerId.toB58String()}`),
+          TEST_PROTOCOL
+        )
+        console.log(`piping msg: ${cmdDef.msg}`)
+
+        await pipe([new TextEncoder().encode(cmdDef.msg)], conn.stream, async (source: Stream['source']) => {
+          for await (const msg of source) {
+            const decoded = new TextDecoder().decode(msg.slice())
+
+            console.log(`Received <${decoded}>`)
```

Comment:
This can use some refactoring to automatically check message reception, e.g. by using a DefferedPromise